TS 3.6 generates broken types references in .d.ts files

<!-- 🚨 STOP 🚨 𝗦𝗧𝗢𝗣 🚨 𝑺𝑻𝑶𝑷 🚨

Half of all issues filed here are duplicates, answered in the FAQ, or not appropriate for the bug tracker. Even if you think you've found a *bug*, please read the FAQ first, especially the Common "Bugs" That Aren't Bugs section!

Please help us by doing the following steps before logging an issue:
  * Search: https://github.com/Microsoft/TypeScript/search?type=Issues
  * Read the FAQ: https://github.com/Microsoft/TypeScript/wiki/FAQ

Please fill in the *entire* template below.
-->

<!-- Please try to reproduce the issue with `typescript@next`. It may have already been fixed. -->
**TypeScript Version:**  3.6.3

<!-- Search terms you tried before logging this (so others can find this issue more easily) -->
**Search Terms:** reference

**Code**

```ts
import { memoize } from 'lodash';

export const fn = memoize((num: number) => num + 1);
```

**Expected behavior:**
The resulting .d.ts file should look like this:
```ts
export declare const fn: ((num: number) => number) & import("lodash").MemoizedFunction;
```

**Actual behavior:**
Instead, since TS 3.6, this is what is generated:
```ts
/// <reference types="lodash/common/common" />
/// <reference types="lodash/common/array" />
/// <reference types="lodash/common/collection" />
/// <reference types="lodash/common/date" />
/// <reference types="lodash/common/function" />
/// <reference types="lodash/common/lang" />
/// <reference types="lodash/common/math" />
/// <reference types="lodash/common/number" />
/// <reference types="lodash/common/object" />
/// <reference types="lodash/common/seq" />
/// <reference types="lodash/common/string" />
/// <reference types="lodash/common/util" />
export declare const fn: ((num: number) => number) & import("lodash").MemoizedFunction;
```
<details>
<summary>This breaks the types for consumers of this .d.ts file:</summary>

```
$ npx tsc
node_modules/@types/lodash/ts3.1/common/common.d.ts:190:15 - error TS2428: All declarations of 'Function' must have identical type parameters.

190     interface Function<T extends (...args: any) => any> extends LoDashImplicitWrapper<T> {
                  ~~~~~~~~

node_modules/@types/lodash/ts3.1/common/common.d.ts:200:15 - error TS2428: All declarations of 'FunctionChain' must have identical type parameters.

200     interface FunctionChain<T extends (...args: any) => any> extends LoDashExplicitWrapper<T> {
                  ~~~~~~~~~~~~~

node_modules/@types/lodash/ts3.1/common/function.d.ts:35:15 - error TS2428: All declarations of 'Function' must have identical type parameters.

35     interface Function<T> {
                 ~~~~~~~~

node_modules/@types/lodash/ts3.1/common/function.d.ts:41:15 - error TS2428: All declarations of 'FunctionChain' must have identical type parameters.

41     interface FunctionChain<T> {
                 ~~~~~~~~~~~~~

node_modules/@types/lodash/ts3.1/common/function.d.ts:297:15 - error TS2428: All declarations of 'Function' must have identical type parameters.

297     interface Function<T> {
                  ~~~~~~~~

node_modules/@types/lodash/ts3.1/common/function.d.ts:309:15 - error TS2428: All declarations of 'FunctionChain' must have identical type parameters.

309     interface FunctionChain<T> {
                  ~~~~~~~~~~~~~

node_modules/@types/lodash/ts3.1/common/function.d.ts:333:15 - error TS2428: All declarations of 'Function' must have identical type parameters.

333     interface Function<T> {
                  ~~~~~~~~

node_modules/@types/lodash/ts3.1/common/function.d.ts:345:15 - error TS2428: All declarations of 'FunctionChain' must have identical type parameters.

345     interface FunctionChain<T> {
                  ~~~~~~~~~~~~~

node_modules/@types/lodash/ts3.1/common/function.d.ts:394:15 - error TS2428: All declarations of 'Function' must have identical type parameters.

394     interface Function<T extends (...args: any) => any> {
                  ~~~~~~~~

node_modules/@types/lodash/ts3.1/common/function.d.ts:400:15 - error TS2428: All declarations of 'FunctionChain' must have identical type parameters.

400     interface FunctionChain<T extends (...args: any) => any> {
                  ~~~~~~~~~~~~~

node_modules/@types/lodash/ts3.1/common/function.d.ts:470:15 - error TS2428: All declarations of 'Function' must have identical type parameters.

470     interface Function<T extends (...args: any) => any> {
                  ~~~~~~~~

node_modules/@types/lodash/ts3.1/common/function.d.ts:476:15 - error TS2428: All declarations of 'FunctionChain' must have identical type parameters.

476     interface FunctionChain<T extends (...args: any) => any> {
                  ~~~~~~~~~~~~~

node_modules/@types/lodash/ts3.1/common/function.d.ts:504:15 - error TS2428: All declarations of 'Function' must have identical type parameters.

504     interface Function<T extends (...args: any) => any> {
                  ~~~~~~~~

node_modules/@types/lodash/ts3.1/common/function.d.ts:510:15 - error TS2428: All declarations of 'FunctionChain' must have identical type parameters.

510     interface FunctionChain<T extends (...args: any) => any> {
                  ~~~~~~~~~~~~~

node_modules/@types/lodash/ts3.1/common/function.d.ts:526:15 - error TS2428: All declarations of 'Function' must have identical type parameters.

526     interface Function<T extends (...args: any) => any> {
                  ~~~~~~~~

node_modules/@types/lodash/ts3.1/common/function.d.ts:530:49 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

530         negate(): Function<(...args: Parameters<T>) => boolean>;
                                                    ~

node_modules/@types/lodash/ts3.1/common/function.d.ts:532:15 - error TS2428: All declarations of 'FunctionChain' must have identical type parameters.

532     interface FunctionChain<T extends (...args: any) => any> {
                  ~~~~~~~~~~~~~

node_modules/@types/lodash/ts3.1/common/function.d.ts:536:54 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

536         negate(): FunctionChain<(...args: Parameters<T>) => boolean>;
                                                         ~

node_modules/@types/lodash/ts3.1/common/function.d.ts:548:15 - error TS2428: All declarations of 'Function' must have identical type parameters.

548     interface Function<T extends (...args: any) => any> {
                  ~~~~~~~~

node_modules/@types/lodash/ts3.1/common/function.d.ts:554:15 - error TS2428: All declarations of 'FunctionChain' must have identical type parameters.

554     interface FunctionChain<T extends (...args: any) => any> {
                  ~~~~~~~~~~~~~

node_modules/@types/lodash/ts3.1/common/function.d.ts:571:15 - error TS2428: All declarations of 'Function' must have identical type parameters.

571     interface Function<T> {
                  ~~~~~~~~

node_modules/@types/lodash/ts3.1/common/function.d.ts:577:15 - error TS2428: All declarations of 'FunctionChain' must have identical type parameters.

577     interface FunctionChain<T> {
                  ~~~~~~~~~~~~~

node_modules/@types/lodash/ts3.1/common/function.d.ts:625:15 - error TS2428: All declarations of 'Function' must have identical type parameters.

625     interface Function<T> {
                  ~~~~~~~~

node_modules/@types/lodash/ts3.1/common/function.d.ts:741:15 - error TS2428: All declarations of 'FunctionChain' must have identical type parameters.

741     interface FunctionChain<T> {
                  ~~~~~~~~~~~~~

node_modules/@types/lodash/ts3.1/common/function.d.ts:902:15 - error TS2428: All declarations of 'Function' must have identical type parameters.

902     interface Function<T> {
                  ~~~~~~~~

node_modules/@types/lodash/ts3.1/common/function.d.ts:1065:15 - error TS2428: All declarations of 'FunctionChain' must have identical type parameters.

1065     interface FunctionChain<T> {
                   ~~~~~~~~~~~~~

node_modules/@types/lodash/ts3.1/common/function.d.ts:1239:15 - error TS2428: All declarations of 'Function' must have identical type parameters.

1239     interface Function<T> {
                   ~~~~~~~~

node_modules/@types/lodash/ts3.1/common/function.d.ts:1245:15 - error TS2428: All declarations of 'FunctionChain' must have identical type parameters.

1245     interface FunctionChain<T> {
                   ~~~~~~~~~~~~~

node_modules/@types/lodash/ts3.1/common/function.d.ts:1264:15 - error TS2428: All declarations of 'Function' must have identical type parameters.

1264     interface Function<T extends (...args: any) => any> {
                   ~~~~~~~~

node_modules/@types/lodash/ts3.1/common/function.d.ts:1270:15 - error TS2428: All declarations of 'FunctionChain' must have identical type parameters.

1270     interface FunctionChain<T extends (...args: any) => any> {
                   ~~~~~~~~~~~~~

node_modules/@types/lodash/ts3.1/common/function.d.ts:1288:15 - error TS2428: All declarations of 'Function' must have identical type parameters.

1288     interface Function<T> {
                   ~~~~~~~~

node_modules/@types/lodash/ts3.1/common/function.d.ts:1292:73 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

1292         spread(start?: number): Function<(...args: any[]) => ReturnType<T>>;
                                                                             ~

node_modules/@types/lodash/ts3.1/common/function.d.ts:1294:15 - error TS2428: All declarations of 'FunctionChain' must have identical type parameters.

1294     interface FunctionChain<T> {
                   ~~~~~~~~~~~~~

node_modules/@types/lodash/ts3.1/common/function.d.ts:1298:78 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

1298         spread(start?: number): FunctionChain<(...args: any[]) => ReturnType<T>>;
                                                                                  ~

node_modules/@types/lodash/ts3.1/common/function.d.ts:1329:15 - error TS2428: All declarations of 'Function' must have identical type parameters.

1329     interface Function<T extends (...args: any) => any> {
                   ~~~~~~~~

node_modules/@types/lodash/ts3.1/common/function.d.ts:1335:15 - error TS2428: All declarations of 'FunctionChain' must have identical type parameters.

1335     interface FunctionChain<T extends (...args: any) => any> {
                   ~~~~~~~~~~~~~

node_modules/@types/lodash/ts3.1/common/function.d.ts:1356:15 - error TS2428: All declarations of 'Function' must have identical type parameters.

1356     interface Function<T> {
                   ~~~~~~~~

node_modules/@types/lodash/ts3.1/common/function.d.ts:1360:34 - error TS2536: Type '"0"' cannot be used to index type 'Parameters<T>'.

1360         unary(): Function<(arg1: Parameters<T>['0']) => ReturnType<T>>;
                                      ~~~~~~~~~~~~~~~~~~

node_modules/@types/lodash/ts3.1/common/function.d.ts:1360:45 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

1360         unary(): Function<(arg1: Parameters<T>['0']) => ReturnType<T>>;
                                                 ~

node_modules/@types/lodash/ts3.1/common/function.d.ts:1360:68 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

1360         unary(): Function<(arg1: Parameters<T>['0']) => ReturnType<T>>;
                                                                        ~

node_modules/@types/lodash/ts3.1/common/function.d.ts:1362:15 - error TS2428: All declarations of 'FunctionChain' must have identical type parameters.

1362     interface FunctionChain<T> {
                   ~~~~~~~~~~~~~

node_modules/@types/lodash/ts3.1/common/function.d.ts:1366:39 - error TS2536: Type '"0"' cannot be used to index type 'Parameters<T>'.

1366         unary(): FunctionChain<(arg1: Parameters<T>['0']) => ReturnType<T>>;
                                           ~~~~~~~~~~~~~~~~~~

node_modules/@types/lodash/ts3.1/common/function.d.ts:1366:50 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

1366         unary(): FunctionChain<(arg1: Parameters<T>['0']) => ReturnType<T>>;
                                                      ~

node_modules/@types/lodash/ts3.1/common/function.d.ts:1366:73 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

1366         unary(): FunctionChain<(arg1: Parameters<T>['0']) => ReturnType<T>>;
                                                                             ~

node_modules/@types/lodash/ts3.1/common/lang.d.ts:33:15 - error TS2428: All declarations of 'Function' must have identical type parameters.

33     interface Function<T extends (...args: any) => any> {
                 ~~~~~~~~

node_modules/@types/lodash/ts3.1/common/lang.d.ts:69:15 - error TS2428: All declarations of 'FunctionChain' must have identical type parameters.

69     interface FunctionChain<T extends (...args: any) => any> {
                 ~~~~~~~~~~~~~

node_modules/@types/lodash/ts3.1/common/seq.d.ts:65:15 - error TS2428: All declarations of 'Function' must have identical type parameters.

65     interface Function<T extends (...args: any) => any> {
                 ~~~~~~~~

node_modules/@types/lodash/ts3.1/common/util.d.ts:205:15 - error TS2428: All declarations of 'Function' must have identical type parameters.

205     interface Function<T extends (...arg: any) => any> {
                  ~~~~~~~~

node_modules/@types/lodash/ts3.1/common/util.d.ts:209:57 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

209         flow<R2, R3, R4, R5, R6, R7>(f2: (a: ReturnType<T>) => R2, f3: (a: R2) => R3, f4: (a: R3) => R4, f5: (a: R4) => R5, f6: (a: R5) => R6, f7: (a: R6) => 
R7): Function<(...args: Parameters<T>) => R7>;
                                                            ~

node_modules/@types/lodash/ts3.1/common/util.d.ts:209:194 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

209         flow<R2, R3, R4, R5, R6, R7>(f2: (a: ReturnType<T>) => R2, f3: (a: R2) => R3, f4: (a: R3) => R4, f5: (a: R4) => R5, f6: (a: R5) => R6, f7: (a: R6) => 
R7): Function<(...args: Parameters<T>) => R7>;
    
                                   ~

node_modules/@types/lodash/ts3.1/common/util.d.ts:213:57 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

213         flow<R2, R3, R4, R5, R6, R7>(f2: (a: ReturnType<T>) => R2, f3: (a: R2) => R3, f4: (a: R3) => R4, f5: (a: R4) => R5, f6: (a: R5) => R6, f7: (a: R6) => 
R7, ...func: Array<Many<(a: any) => any>>): Function<(...args: Parameters<T>) => any>;
                                                            ~

node_modules/@types/lodash/ts3.1/common/util.d.ts:213:233 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

213         flow<R2, R3, R4, R5, R6, R7>(f2: (a: ReturnType<T>) => R2, f3: (a: R2) => R3, f4: (a: R3) => R4, f5: (a: R4) => R5, f6: (a: R5) => R6, f7: (a: R6) => 
R7, ...func: Array<Many<(a: any) => any>>): Function<(...args: Parameters<T>) => any>;
    
                                                                          ~

node_modules/@types/lodash/ts3.1/common/util.d.ts:217:53 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

217         flow<R2, R3, R4, R5, R6>(f2: (a: ReturnType<T>) => R2, f3: (a: R2) => R3, f4: (a: R3) => R4, f5: (a: R4) => R5, f6: (a: R5) => R6): Function<(...args: Parameters<T>) => R6>;
                                                        ~

node_modules/@types/lodash/ts3.1/common/util.d.ts:217:171 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

217         flow<R2, R3, R4, R5, R6>(f2: (a: ReturnType<T>) => R2, f3: (a: R2) => R3, f4: (a: R3) => R4, f5: (a: R4) => R5, f6: (a: R5) => R6): Function<(...args: Parameters<T>) => R6>;
    
            ~

node_modules/@types/lodash/ts3.1/common/util.d.ts:221:49 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

221         flow<R2, R3, R4, R5>(f2: (a: ReturnType<T>) => R2, f3: (a: R2) => R3, f4: (a: R3) => R4, f5: (a: R4) => R5): Function<(...args: Parameters<T>) => R5>;                                                    ~

node_modules/@types/lodash/ts3.1/common/util.d.ts:221:148 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

221         flow<R2, R3, R4, R5>(f2: (a: ReturnType<T>) => R2, f3: (a: R2) => R3, f4: (a: R3) => R4, f5: (a: R4) => R5): Function<(...args: Parameters<T>) => R5>;                                                                                                                                                       ~

node_modules/@types/lodash/ts3.1/common/util.d.ts:225:45 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

225         flow<R2, R3, R4>(f2: (a: ReturnType<T>) => R2, f3: (a: R2) => R3, f4: (a: R3) => R4): Function<(...args: Parameters<T>) => R4>;
                                                ~

node_modules/@types/lodash/ts3.1/common/util.d.ts:225:125 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

225         flow<R2, R3, R4>(f2: (a: ReturnType<T>) => R2, f3: (a: R2) => R3, f4: (a: R3) => R4): Function<(...args: Parameters<T>) => R4>;
                                                                                                                                ~

node_modules/@types/lodash/ts3.1/common/util.d.ts:229:41 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

229         flow<R2, R3>(f2: (a: ReturnType<T>) => R2, f3: (a: R2) => R3): Function<(...args: Parameters<T>) => R3>;
                                            ~

node_modules/@types/lodash/ts3.1/common/util.d.ts:229:102 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

229         flow<R2, R3>(f2: (a: ReturnType<T>) => R2, f3: (a: R2) => R3): Function<(...args: Parameters<T>) => R3>;
                                                                                                         ~

node_modules/@types/lodash/ts3.1/common/util.d.ts:233:37 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

233         flow<R2>(f2: (a: ReturnType<T>) => R2): Function<(...args: Parameters<T>) => R2>;
                                        ~

node_modules/@types/lodash/ts3.1/common/util.d.ts:233:79 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

233         flow<R2>(f2: (a: ReturnType<T>) => R2): Function<(...args: Parameters<T>) => R2>;
                                                                                  ~

node_modules/@types/lodash/ts3.1/common/util.d.ts:239:15 - error TS2428: All declarations of 'FunctionChain' must have identical type parameters.

239     interface FunctionChain<T> {
                  ~~~~~~~~~~~~~

node_modules/@types/lodash/ts3.1/common/util.d.ts:243:57 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

243         flow<R2, R3, R4, R5, R6, R7>(f2: (a: ReturnType<T>) => R2, f3: (a: R2) => R3, f4: (a: R3) => R4, f5: (a: R4) => R5, f6: (a: R5) => R6, f7: (a: R6) => 
R7): FunctionChain<(...args: Parameters<T>) => R7>;
                                                            ~

node_modules/@types/lodash/ts3.1/common/util.d.ts:243:199 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

243         flow<R2, R3, R4, R5, R6, R7>(f2: (a: ReturnType<T>) => R2, f3: (a: R2) => R3, f4: (a: R3) => R4, f5: (a: R4) => R5, f6: (a: R5) => R6, f7: (a: R6) => 
R7): FunctionChain<(...args: Parameters<T>) => R7>;
    
                                        ~

node_modules/@types/lodash/ts3.1/common/util.d.ts:247:57 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

247         flow<R2, R3, R4, R5, R6, R7>(f2: (a: ReturnType<T>) => R2, f3: (a: R2) => R3, f4: (a: R3) => R4, f5: (a: R4) => R5, f6: (a: R5) => R6, f7: (a: R6) => 
R7, ...func: Array<Many<(a: any) => any>>): FunctionChain<(...args: Parameters<T>) => any>;
                                                            ~

node_modules/@types/lodash/ts3.1/common/util.d.ts:247:238 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

247         flow<R2, R3, R4, R5, R6, R7>(f2: (a: ReturnType<T>) => R2, f3: (a: R2) => R3, f4: (a: R3) => R4, f5: (a: R4) => R5, f6: (a: R5) => R6, f7: (a: R6) => 
R7, ...func: Array<Many<(a: any) => any>>): FunctionChain<(...args: Parameters<T>) => any>;
    
                                                                               ~

node_modules/@types/lodash/ts3.1/common/util.d.ts:251:53 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

251         flow<R2, R3, R4, R5, R6>(f2: (a: ReturnType<T>) => R2, f3: (a: R2) => R3, f4: (a: R3) => R4, f5: (a: R4) => R5, f6: (a: R5) => R6): FunctionChain<(...args: Parameters<T>) => R6>;
                                                        ~

node_modules/@types/lodash/ts3.1/common/util.d.ts:251:176 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

251         flow<R2, R3, R4, R5, R6>(f2: (a: ReturnType<T>) => R2, f3: (a: R2) => R3, f4: (a: R3) => R4, f5: (a: R4) => R5, f6: (a: R5) => R6): FunctionChain<(...args: Parameters<T>) => R6>;
    
                 ~

node_modules/@types/lodash/ts3.1/common/util.d.ts:255:49 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

255         flow<R2, R3, R4, R5>(f2: (a: ReturnType<T>) => R2, f3: (a: R2) => R3, f4: (a: R3) => R4, f5: (a: R4) => R5): FunctionChain<(...args: Parameters<T>) => R5>;
                                                    ~

node_modules/@types/lodash/ts3.1/common/util.d.ts:255:153 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

255         flow<R2, R3, R4, R5>(f2: (a: ReturnType<T>) => R2, f3: (a: R2) => R3, f4: (a: R3) => R4, f5: (a: R4) => R5): FunctionChain<(...args: Parameters<T>) => R5>;
                                                                                                                                                            ~     

node_modules/@types/lodash/ts3.1/common/util.d.ts:259:45 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

259         flow<R2, R3, R4>(f2: (a: ReturnType<T>) => R2, f3: (a: R2) => R3, f4: (a: R3) => R4): FunctionChain<(...args: Parameters<T>) => R4>;
                                                ~

node_modules/@types/lodash/ts3.1/common/util.d.ts:259:130 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

259         flow<R2, R3, R4>(f2: (a: ReturnType<T>) => R2, f3: (a: R2) => R3, f4: (a: R3) => R4): FunctionChain<(...args: Parameters<T>) => R4>;
                                                                                                                                     ~

node_modules/@types/lodash/ts3.1/common/util.d.ts:263:41 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

263         flow<R2, R3>(f2: (a: ReturnType<T>) => R2, f3: (a: R2) => R3): FunctionChain<(...args: Parameters<T>) => R3>;
                                            ~

node_modules/@types/lodash/ts3.1/common/util.d.ts:263:107 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

263         flow<R2, R3>(f2: (a: ReturnType<T>) => R2, f3: (a: R2) => R3): FunctionChain<(...args: Parameters<T>) => R3>;
                                                                                                              ~

node_modules/@types/lodash/ts3.1/common/util.d.ts:267:37 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

267         flow<R2>(f2: (a: ReturnType<T>) => R2): FunctionChain<(...args: Parameters<T>) => R2>;
                                        ~

node_modules/@types/lodash/ts3.1/common/util.d.ts:267:84 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

267         flow<R2>(f2: (a: ReturnType<T>) => R2): FunctionChain<(...args: Parameters<T>) => R2>;
                                                                                       ~

node_modules/@types/lodash/ts3.1/common/util.d.ts:308:15 - error TS2428: All declarations of 'Function' must have identical type parameters.

308     interface Function<T> {
                  ~~~~~~~~

node_modules/@types/lodash/ts3.1/common/util.d.ts:312:71 - error TS2536: Type '"0"' cannot be used to index type 'Parameters<T>'.

312         flowRight<A extends any[], R1, R2, R3, R4, R5>(f6: (a: R5) => Parameters<T>["0"], f5: (a: R4) => R5, f4: (a: R3) => R4, f3: (a: R2) => R3, f2: (a: R1) => R2, f1: (...args: A) => R1): Function<(...args: A) => ReturnType<T>>;
                                                                          ~~~~~~~~~~~~~~~~~~

node_modules/@types/lodash/ts3.1/common/util.d.ts:312:82 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

312         flowRight<A extends any[], R1, R2, R3, R4, R5>(f6: (a: R5) => Parameters<T>["0"], f5: (a: R4) => R5, f4: (a: R3) => R4, f3: (a: R2) => R3, f2: (a: R1) => R2, f1: (...args: A) => R1): Function<(...args: A) => ReturnType<T>>;
                                                                                     ~

node_modules/@types/lodash/ts3.1/common/util.d.ts:312:228 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

312         flowRight<A extends any[], R1, R2, R3, R4, R5>(f6: (a: R5) => Parameters<T>["0"], f5: (a: R4) => R5, f4: (a: R3) => R4, f3: (a: R2) => R3, f2: (a: R1) => R2, f1: (...args: A) => R1): Function<(...args: A) => ReturnType<T>>;
    
                                                                     ~

node_modules/@types/lodash/ts3.1/common/util.d.ts:316:67 - error TS2536: Type '"0"' cannot be used to index type 'Parameters<T>'.

316         flowRight<A extends any[], R1, R2, R3, R4>(f5: (a: R4) => Parameters<T>["0"], f4: (a: R3) => R4, f3: (a: R2) => R3, f2: (a: R1) => R2, f1: (...args: A) => R1): Function<(...args: A) => ReturnType<T>>;
                                                                      ~~~~~~~~~~~~~~~~~~

node_modules/@types/lodash/ts3.1/common/util.d.ts:316:78 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

316         flowRight<A extends any[], R1, R2, R3, R4>(f5: (a: R4) => Parameters<T>["0"], f4: (a: R3) => R4, f3: (a: R2) => R3, f2: (a: R1) => R2, f1: (...args: A) => R1): Function<(...args: A) => ReturnType<T>>;
                                                                                 ~

node_modules/@types/lodash/ts3.1/common/util.d.ts:316:205 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

316         flowRight<A extends any[], R1, R2, R3, R4>(f5: (a: R4) => Parameters<T>["0"], f4: (a: R3) => R4, f3: (a: R2) => R3, f2: (a: R1) => R2, f1: (...args: A) => R1): Function<(...args: A) => ReturnType<T>>;
    
                                              ~

node_modules/@types/lodash/ts3.1/common/util.d.ts:320:63 - error TS2536: Type '"0"' cannot be used to index type 'Parameters<T>'.

320         flowRight<A extends any[], R1, R2, R3>(f4: (a: R3) => Parameters<T>["0"], f3: (a: R2) => R3, f2: (a: R1) => R2, f1: (...args: A) => R1): Function<(...args: A) => ReturnType<T>>;
                                                                  ~~~~~~~~~~~~~~~~~~

node_modules/@types/lodash/ts3.1/common/util.d.ts:320:74 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

320         flowRight<A extends any[], R1, R2, R3>(f4: (a: R3) => Parameters<T>["0"], f3: (a: R2) => R3, f2: (a: R1) => R2, f1: (...args: A) => R1): Function<(...args: A) => ReturnType<T>>;
                                                                             ~

node_modules/@types/lodash/ts3.1/common/util.d.ts:320:182 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

320         flowRight<A extends any[], R1, R2, R3>(f4: (a: R3) => Parameters<T>["0"], f3: (a: R2) => R3, f2: (a: R1) => R2, f1: (...args: A) => R1): Function<(...args: A) => ReturnType<T>>;
    
                       ~

node_modules/@types/lodash/ts3.1/common/util.d.ts:324:59 - error TS2536: Type '"0"' cannot be used to index type 'Parameters<T>'.

324         flowRight<A extends any[], R1, R2>(f3: (a: R2) => Parameters<T>["0"], f2: (a: R1) => R2, f1: (...args: A) => R1): Function<(...args: A) => ReturnType<T>>;
                                                              ~~~~~~~~~~~~~~~~~~

node_modules/@types/lodash/ts3.1/common/util.d.ts:324:70 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

324         flowRight<A extends any[], R1, R2>(f3: (a: R2) => Parameters<T>["0"], f2: (a: R1) => R2, f1: (...args: A) => R1): Function<(...args: A) => ReturnType<T>>;
                                                                         ~

node_modules/@types/lodash/ts3.1/common/util.d.ts:324:159 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

324         flowRight<A extends any[], R1, R2>(f3: (a: R2) => Parameters<T>["0"], f2: (a: R1) => R2, f1: (...args: A) => R1): Function<(...args: A) => ReturnType<T>>;
    
~

node_modules/@types/lodash/ts3.1/common/util.d.ts:328:55 - error TS2536: Type '"0"' cannot be used to index type 'Parameters<T>'.

328         flowRight<A extends any[], R1>(f2: (a: R1) => Parameters<T>["0"], f1: (...args: A) => R1): Function<(...args: A) => ReturnType<T>>;
                                                          ~~~~~~~~~~~~~~~~~~

node_modules/@types/lodash/ts3.1/common/util.d.ts:328:66 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

328         flowRight<A extends any[], R1>(f2: (a: R1) => Parameters<T>["0"], f1: (...args: A) => R1): Function<(...args: A) => ReturnType<T>>;
                                                                     ~

node_modules/@types/lodash/ts3.1/common/util.d.ts:328:136 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

328         flowRight<A extends any[], R1>(f2: (a: R1) => Parameters<T>["0"], f1: (...args: A) => R1): Function<(...args: A) => ReturnType<T>>;
                                                                                                                                           ~

node_modules/@types/lodash/ts3.1/common/util.d.ts:332:56 - error TS2536: Type '"0"' cannot be used to index type 'Parameters<T>'.

332         flowRight<A extends any[]>(f1: (...args: A) => Parameters<T>["0"]): Function<(...args: A) => ReturnType<T>>;
                                                           ~~~~~~~~~~~~~~~~~~

node_modules/@types/lodash/ts3.1/common/util.d.ts:332:67 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

332         flowRight<A extends any[]>(f1: (...args: A) => Parameters<T>["0"]): Function<(...args: A) => ReturnType<T>>;
                                                                      ~

node_modules/@types/lodash/ts3.1/common/util.d.ts:332:113 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

332         flowRight<A extends any[]>(f1: (...args: A) => Parameters<T>["0"]): Function<(...args: A) => ReturnType<T>>;
                                                                                                                    ~

node_modules/@types/lodash/ts3.1/common/util.d.ts:338:15 - error TS2428: All declarations of 'FunctionChain' must have identical type parameters.

338     interface FunctionChain<T> {
                  ~~~~~~~~~~~~~

node_modules/@types/lodash/ts3.1/common/util.d.ts:342:71 - error TS2536: Type '"0"' cannot be used to index type 'Parameters<T>'.

342         flowRight<A extends any[], R1, R2, R3, R4, R5>(f6: (a: R5) => Parameters<T>["0"], f5: (a: R4) => R5, f4: (a: R3) => R4, f3: (a: R2) => R3, f2: (a: R1) => R2, f1: (...args: A) => R1): FunctionChain<(...args: A) => ReturnType<T>>;
                                                                          ~~~~~~~~~~~~~~~~~~

node_modules/@types/lodash/ts3.1/common/util.d.ts:342:82 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

342         flowRight<A extends any[], R1, R2, R3, R4, R5>(f6: (a: R5) => Parameters<T>["0"], f5: (a: R4) => R5, f4: (a: R3) => R4, f3: (a: R2) => R3, f2: (a: R1) => R2, f1: (...args: A) => R1): FunctionChain<(...args: A) => ReturnType<T>>;
                                                                                     ~

node_modules/@types/lodash/ts3.1/common/util.d.ts:342:233 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

342         flowRight<A extends any[], R1, R2, R3, R4, R5>(f6: (a: R5) => Parameters<T>["0"], f5: (a: R4) => R5, f4: (a: R3) => R4, f3: (a: R2) => R3, f2: (a: R1) => R2, f1: (...args: A) => R1): FunctionChain<(...args: A) => ReturnType<T>>;
    
                                                                          ~

node_modules/@types/lodash/ts3.1/common/util.d.ts:346:67 - error TS2536: Type '"0"' cannot be used to index type 'Parameters<T>'.

346         flowRight<A extends any[], R1, R2, R3, R4>(f5: (a: R4) => Parameters<T>["0"], f4: (a: R3) => R4, f3: (a: R2) => R3, f2: (a: R1) => R2, f1: (...args: A) => R1): FunctionChain<(...args: A) => ReturnType<T>>;
                                                                      ~~~~~~~~~~~~~~~~~~

node_modules/@types/lodash/ts3.1/common/util.d.ts:346:78 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

346         flowRight<A extends any[], R1, R2, R3, R4>(f5: (a: R4) => Parameters<T>["0"], f4: (a: R3) => R4, f3: (a: R2) => R3, f2: (a: R1) => R2, f1: (...args: A) => R1): FunctionChain<(...args: A) => ReturnType<T>>;
                                                                                 ~

node_modules/@types/lodash/ts3.1/common/util.d.ts:346:210 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

346         flowRight<A extends any[], R1, R2, R3, R4>(f5: (a: R4) => Parameters<T>["0"], f4: (a: R3) => R4, f3: (a: R2) => R3, f2: (a: R1) => R2, f1: (...args: A) => R1): FunctionChain<(...args: A) => ReturnType<T>>;
    
                                                   ~

node_modules/@types/lodash/ts3.1/common/util.d.ts:350:63 - error TS2536: Type '"0"' cannot be used to index type 'Parameters<T>'.

350         flowRight<A extends any[], R1, R2, R3>(f4: (a: R3) => Parameters<T>["0"], f3: (a: R2) => R3, f2: (a: R1) => R2, f1: (...args: A) => R1): FunctionChain<(...args: A) => ReturnType<T>>;
                                                                  ~~~~~~~~~~~~~~~~~~

node_modules/@types/lodash/ts3.1/common/util.d.ts:350:74 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

350         flowRight<A extends any[], R1, R2, R3>(f4: (a: R3) => Parameters<T>["0"], f3: (a: R2) => R3, f2: (a: R1) => R2, f1: (...args: A) => R1): FunctionChain<(...args: A) => ReturnType<T>>;
                                                                             ~

node_modules/@types/lodash/ts3.1/common/util.d.ts:350:187 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

350         flowRight<A extends any[], R1, R2, R3>(f4: (a: R3) => Parameters<T>["0"], f3: (a: R2) => R3, f2: (a: R1) => R2, f1: (...args: A) => R1): FunctionChain<(...args: A) => ReturnType<T>>;
    
                            ~

node_modules/@types/lodash/ts3.1/common/util.d.ts:354:59 - error TS2536: Type '"0"' cannot be used to index type 'Parameters<T>'.

354         flowRight<A extends any[], R1, R2>(f3: (a: R2) => Parameters<T>["0"], f2: (a: R1) => R2, f1: (...args: A) => R1): FunctionChain<(...args: A) => ReturnType<T>>;
                                                              ~~~~~~~~~~~~~~~~~~

node_modules/@types/lodash/ts3.1/common/util.d.ts:354:70 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

354         flowRight<A extends any[], R1, R2>(f3: (a: R2) => Parameters<T>["0"], f2: (a: R1) => R2, f1: (...args: A) => R1): FunctionChain<(...args: A) => ReturnType<T>>;
                                                                         ~

node_modules/@types/lodash/ts3.1/common/util.d.ts:354:164 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

354         flowRight<A extends any[], R1, R2>(f3: (a: R2) => Parameters<T>["0"], f2: (a: R1) => R2, f1: (...args: A) => R1): FunctionChain<(...args: A) => ReturnType<T>>;
    
     ~

node_modules/@types/lodash/ts3.1/common/util.d.ts:358:55 - error TS2536: Type '"0"' cannot be used to index type 'Parameters<T>'.

358         flowRight<A extends any[], R1>(f2: (a: R1) => Parameters<T>["0"], f1: (...args: A) => R1): FunctionChain<(...args: A) => ReturnType<T>>;
                                                          ~~~~~~~~~~~~~~~~~~

node_modules/@types/lodash/ts3.1/common/util.d.ts:358:66 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

358         flowRight<A extends any[], R1>(f2: (a: R1) => Parameters<T>["0"], f1: (...args: A) => R1): FunctionChain<(...args: A) => ReturnType<T>>;
                                                                     ~

node_modules/@types/lodash/ts3.1/common/util.d.ts:358:141 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

358         flowRight<A extends any[], R1>(f2: (a: R1) => Parameters<T>["0"], f1: (...args: A) => R1): FunctionChain<(...args: A) => ReturnType<T>>;
                                                                                                                                                ~

node_modules/@types/lodash/ts3.1/common/util.d.ts:362:56 - error TS2536: Type '"0"' cannot be used to index type 'Parameters<T>'.

362         flowRight<A extends any[]>(f1: (...args: A) => Parameters<T>["0"]): FunctionChain<(...args: A) => ReturnType<T>>;
                                                           ~~~~~~~~~~~~~~~~~~

node_modules/@types/lodash/ts3.1/common/util.d.ts:362:67 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

362         flowRight<A extends any[]>(f1: (...args: A) => Parameters<T>["0"]): FunctionChain<(...args: A) => ReturnType<T>>;
                                                                      ~

node_modules/@types/lodash/ts3.1/common/util.d.ts:362:118 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

362         flowRight<A extends any[]>(f1: (...args: A) => Parameters<T>["0"]): FunctionChain<(...args: A) => ReturnType<T>>;
                                                                                                                         ~

node_modules/@types/lodash/ts3.1/common/util.d.ts:429:15 - error TS2428: All declarations of 'Function' must have identical type parameters.

429     interface Function<T extends (...args: any) => any> {
                  ~~~~~~~~

node_modules/@types/lodash/ts3.1/common/util.d.ts:453:15 - error TS2428: All declarations of 'FunctionChain' must have identical type parameters.

453     interface FunctionChain<T extends (...args: any) => any> {
                  ~~~~~~~~~~~~~

node_modules/@types/lodash/ts3.1/common/util.d.ts:728:15 - error TS2428: All declarations of 'Function' must have identical type parameters.

728     interface Function<T> {
                  ~~~~~~~~

node_modules/@types/lodash/ts3.1/common/util.d.ts:732:126 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

732         over<TResult>(...iteratees: Array<Many<(...args: any[]) => TResult>>): Function<(...args: any[]) => Array<ReturnType<T> | TResult>>;
                                                                                                                                 ~

node_modules/@types/lodash/ts3.1/common/util.d.ts:740:15 - error TS2428: All declarations of 'FunctionChain' must have identical type parameters.

740     interface FunctionChain<T> {
                  ~~~~~~~~~~~~~

node_modules/@types/lodash/ts3.1/common/util.d.ts:744:131 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

744         over<TResult>(...iteratees: Array<Many<(...args: any[]) => TResult>>): FunctionChain<(...args: any[]) => Array<ReturnType<T> | TResult>>;
                                                                                                                                      ~

node_modules/@types/lodash/ts3.1/common/util.d.ts:763:15 - error TS2428: All declarations of 'Function' must have identical type parameters.

763     interface Function<T> {
                  ~~~~~~~~

node_modules/@types/lodash/ts3.1/common/util.d.ts:767:115 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

767         overEvery<TArgs>(...iteratees: Array<Many<(...args: TArgs[]) => boolean>>): Function<(...args: Parameters<T> | TArgs[]) => boolean>;
                                                                                                                      ~

node_modules/@types/lodash/ts3.1/common/util.d.ts:775:15 - error TS2428: All declarations of 'FunctionChain' must have identical type parameters.

775     interface FunctionChain<T> {
                  ~~~~~~~~~~~~~

node_modules/@types/lodash/ts3.1/common/util.d.ts:779:120 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

779         overEvery<TArgs>(...iteratees: Array<Many<(...args: TArgs[]) => boolean>>): FunctionChain<(...args: Parameters<T> | TArgs[]) => boolean>;
                                                                                                                           ~

node_modules/@types/lodash/ts3.1/common/util.d.ts:798:15 - error TS2428: All declarations of 'Function' must have identical type parameters.

798     interface Function<T> {
                  ~~~~~~~~

node_modules/@types/lodash/ts3.1/common/util.d.ts:802:114 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

802         overSome<TArgs>(...iteratees: Array<Many<(...args: TArgs[]) => boolean>>): Function<(...args: Parameters<T> | TArgs[]) => boolean>;
                                                                                                                     ~

node_modules/@types/lodash/ts3.1/common/util.d.ts:810:15 - error TS2428: All declarations of 'FunctionChain' must have identical type parameters.

810     interface FunctionChain<T> {
                  ~~~~~~~~~~~~~

node_modules/@types/lodash/ts3.1/common/util.d.ts:814:119 - error TS2344: Type 'T' does not satisfy the constraint '(...args: any) => any'.

814         overSome<TArgs>(...iteratees: Array<Many<(...args: TArgs[]) => boolean>>): FunctionChain<(...args: Parameters<T> | TArgs[]) => boolean>;
                                                                                                                          ~


Found 129 errors.
```


</details>

**Playground Link:** <!-- A link to a TypeScript Playground "Share" link which demonstrates this behavior --> The playground doesn't have TS 3.6 yet. ¯\\\_(ツ)_/¯

**Related Issues:** <!-- Did you find other bugs that looked similar? --> I don't think so.

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -8971,7 +8971,7 @@ namespace ts {
             return undefined;
         }
 
-        function getConstraintDeclaration(type: TypeParameter) {
+        function getConstraintDeclaration(type: TypeParameter): TypeNode | undefined {
             return mapDefined(filter(type.symbol && type.symbol.declarations, isTypeParameterDeclaration), getEffectiveConstraintOfTypeParameter)[0];
         }
 

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -8972,8 +8972,7 @@ namespace ts {
         }
 
         function getConstraintDeclaration(type: TypeParameter) {
-            const decl = type.symbol && getDeclarationOfKind<TypeParameterDeclaration>(type.symbol, SyntaxKind.TypeParameter);
-            return decl && getEffectiveConstraintOfTypeParameter(decl);
+            return mapDefined(filter(type.symbol && type.symbol.declarations, isTypeParameterDeclaration), getEffectiveConstraintOfTypeParameter)[0];
         }
 
         function getInferredTypeParameterConstraint(typeParameter: TypeParameter) {
@@ -29233,11 +29232,10 @@ namespace ts {
                     const constraint = getEffectiveConstraintOfTypeParameter(source);
                     const sourceConstraint = constraint && getTypeFromTypeNode(constraint);
                     const targetConstraint = getConstraintOfTypeParameter(target);
-                    if (sourceConstraint) {
-                        // relax check if later interface augmentation has no constraint
-                        if (!targetConstraint || !isTypeIdenticalTo(sourceConstraint, targetConstraint)) {
-                            return false;
-                        }
+                    // relax check if later interface augmentation has no constraint, it's more broad and is OK to merge with
+                    // a more constrained interface (this could be generalized to a full heirarchy check, but that's maybe overkill)
+                    if (sourceConstraint && targetConstraint && !isTypeIdenticalTo(sourceConstraint, targetConstraint)) {
+                        return false;
                     }
 
                     // If the type parameter node has a default and it is not identical to the default

--- a/tests/baselines/reference/interfaceMergedUnconstrainedNoErrorIrrespectiveOfOrder.js
+++ b/tests/baselines/reference/interfaceMergedUnconstrainedNoErrorIrrespectiveOfOrder.js
@@ -1,0 +1,28 @@
+//// [tests/cases/compiler/interfaceMergedUnconstrainedNoErrorIrrespectiveOfOrder.ts] ////
+
+//// [working.ts]
+// minmal samples from #33395
+export namespace ns {
+    interface Function<T extends (...args: any) => any> {
+        throttle(): Function<T>;
+    }
+    interface Function<T> {
+        unary(): Function<() => ReturnType<T>>;
+    }
+}
+//// [regression.ts]
+export namespace ns {
+    interface Function<T> {
+        unary(): Function<() => ReturnType<T>>;
+    }
+    interface Function<T extends (...args: any) => any> {
+        throttle(): Function<T>;
+    }
+}
+
+//// [working.js]
+"use strict";
+exports.__esModule = true;
+//// [regression.js]
+"use strict";
+exports.__esModule = true;

--- a/tests/baselines/reference/interfaceMergedUnconstrainedNoErrorIrrespectiveOfOrder.symbols
+++ b/tests/baselines/reference/interfaceMergedUnconstrainedNoErrorIrrespectiveOfOrder.symbols
@@ -1,0 +1,51 @@
+=== tests/cases/compiler/working.ts ===
+// minmal samples from #33395
+export namespace ns {
+>ns : Symbol(ns, Decl(working.ts, 0, 0))
+
+    interface Function<T extends (...args: any) => any> {
+>Function : Symbol(Function, Decl(working.ts, 1, 21), Decl(working.ts, 4, 5))
+>T : Symbol(T, Decl(working.ts, 2, 23), Decl(working.ts, 5, 23))
+>args : Symbol(args, Decl(working.ts, 2, 34))
+
+        throttle(): Function<T>;
+>throttle : Symbol(Function.throttle, Decl(working.ts, 2, 57))
+>Function : Symbol(Function, Decl(working.ts, 1, 21), Decl(working.ts, 4, 5))
+>T : Symbol(T, Decl(working.ts, 2, 23), Decl(working.ts, 5, 23))
+    }
+    interface Function<T> {
+>Function : Symbol(Function, Decl(working.ts, 1, 21), Decl(working.ts, 4, 5))
+>T : Symbol(T, Decl(working.ts, 2, 23), Decl(working.ts, 5, 23))
+
+        unary(): Function<() => ReturnType<T>>;
+>unary : Symbol(Function.unary, Decl(working.ts, 5, 27))
+>Function : Symbol(Function, Decl(working.ts, 1, 21), Decl(working.ts, 4, 5))
+>ReturnType : Symbol(ReturnType, Decl(lib.es5.d.ts, --, --))
+>T : Symbol(T, Decl(working.ts, 2, 23), Decl(working.ts, 5, 23))
+    }
+}
+=== tests/cases/compiler/regression.ts ===
+export namespace ns {
+>ns : Symbol(ns, Decl(regression.ts, 0, 0))
+
+    interface Function<T> {
+>Function : Symbol(Function, Decl(regression.ts, 0, 21), Decl(regression.ts, 3, 5))
+>T : Symbol(T, Decl(regression.ts, 1, 23), Decl(regression.ts, 4, 23))
+
+        unary(): Function<() => ReturnType<T>>;
+>unary : Symbol(Function.unary, Decl(regression.ts, 1, 27))
+>Function : Symbol(Function, Decl(regression.ts, 0, 21), Decl(regression.ts, 3, 5))
+>ReturnType : Symbol(ReturnType, Decl(lib.es5.d.ts, --, --))
+>T : Symbol(T, Decl(regression.ts, 1, 23), Decl(regression.ts, 4, 23))
+    }
+    interface Function<T extends (...args: any) => any> {
+>Function : Symbol(Function, Decl(regression.ts, 0, 21), Decl(regression.ts, 3, 5))
+>T : Symbol(T, Decl(regression.ts, 1, 23), Decl(regression.ts, 4, 23))
+>args : Symbol(args, Decl(regression.ts, 4, 34))
+
+        throttle(): Function<T>;
+>throttle : Symbol(Function.throttle, Decl(regression.ts, 4, 57))
+>Function : Symbol(Function, Decl(regression.ts, 0, 21), Decl(regression.ts, 3, 5))
+>T : Symbol(T, Decl(regression.ts, 1, 23), Decl(regression.ts, 4, 23))
+    }
+}

--- a/tests/baselines/reference/interfaceMergedUnconstrainedNoErrorIrrespectiveOfOrder.types
+++ b/tests/baselines/reference/interfaceMergedUnconstrainedNoErrorIrrespectiveOfOrder.types
@@ -1,0 +1,27 @@
+=== tests/cases/compiler/working.ts ===
+// minmal samples from #33395
+export namespace ns {
+    interface Function<T extends (...args: any) => any> {
+>args : any
+
+        throttle(): Function<T>;
+>throttle : () => Function<T>
+    }
+    interface Function<T> {
+        unary(): Function<() => ReturnType<T>>;
+>unary : () => Function<() => ReturnType<T>>
+    }
+}
+=== tests/cases/compiler/regression.ts ===
+export namespace ns {
+    interface Function<T> {
+        unary(): Function<() => ReturnType<T>>;
+>unary : () => Function<() => ReturnType<T>>
+    }
+    interface Function<T extends (...args: any) => any> {
+>args : any
+
+        throttle(): Function<T>;
+>throttle : () => Function<T>
+    }
+}

--- a/tests/baselines/reference/twoGenericInterfacesWithDifferentConstraints.errors.txt
+++ b/tests/baselines/reference/twoGenericInterfacesWithDifferentConstraints.errors.txt
@@ -75,7 +75,7 @@ tests/cases/conformance/interfaces/declarationMerging/twoGenericInterfacesWithDi
       x: T;
     }
     
-    interface C<T extends number> { // error
+    interface C<T extends number> { // ok
       y: T;
     }
     

--- a/tests/baselines/reference/twoGenericInterfacesWithDifferentConstraints.errors.txt
+++ b/tests/baselines/reference/twoGenericInterfacesWithDifferentConstraints.errors.txt
@@ -4,11 +4,9 @@ tests/cases/conformance/interfaces/declarationMerging/twoGenericInterfacesWithDi
 tests/cases/conformance/interfaces/declarationMerging/twoGenericInterfacesWithDifferentConstraints.ts(14,15): error TS2428: All declarations of 'B' must have identical type parameters.
 tests/cases/conformance/interfaces/declarationMerging/twoGenericInterfacesWithDifferentConstraints.ts(32,22): error TS2428: All declarations of 'A' must have identical type parameters.
 tests/cases/conformance/interfaces/declarationMerging/twoGenericInterfacesWithDifferentConstraints.ts(38,22): error TS2428: All declarations of 'A' must have identical type parameters.
-tests/cases/conformance/interfaces/declarationMerging/twoGenericInterfacesWithDifferentConstraints.ts(53,11): error TS2428: All declarations of 'C' must have identical type parameters.
-tests/cases/conformance/interfaces/declarationMerging/twoGenericInterfacesWithDifferentConstraints.ts(57,11): error TS2428: All declarations of 'C' must have identical type parameters.
 
 
-==== tests/cases/conformance/interfaces/declarationMerging/twoGenericInterfacesWithDifferentConstraints.ts (8 errors) ====
+==== tests/cases/conformance/interfaces/declarationMerging/twoGenericInterfacesWithDifferentConstraints.ts (6 errors) ====
     interface A<T extends Date> {
               ~
 !!! error TS2428: All declarations of 'A' must have identical type parameters.
@@ -74,14 +72,10 @@ tests/cases/conformance/interfaces/declarationMerging/twoGenericInterfacesWithDi
     }
     
     interface C<T> {
-              ~
-!!! error TS2428: All declarations of 'C' must have identical type parameters.
       x: T;
     }
     
     interface C<T extends number> { // error
-              ~
-!!! error TS2428: All declarations of 'C' must have identical type parameters.
       y: T;
     }
     

--- a/tests/baselines/reference/twoGenericInterfacesWithDifferentConstraints.js
+++ b/tests/baselines/reference/twoGenericInterfacesWithDifferentConstraints.js
@@ -55,7 +55,7 @@ interface C<T> {
   x: T;
 }
 
-interface C<T extends number> { // error
+interface C<T extends number> { // ok
   y: T;
 }
 

--- a/tests/baselines/reference/twoGenericInterfacesWithDifferentConstraints.symbols
+++ b/tests/baselines/reference/twoGenericInterfacesWithDifferentConstraints.symbols
@@ -137,7 +137,7 @@ interface C<T> {
 >T : Symbol(T, Decl(twoGenericInterfacesWithDifferentConstraints.ts, 52, 12), Decl(twoGenericInterfacesWithDifferentConstraints.ts, 56, 12))
 }
 
-interface C<T extends number> { // error
+interface C<T extends number> { // ok
 >C : Symbol(C, Decl(twoGenericInterfacesWithDifferentConstraints.ts, 50, 1), Decl(twoGenericInterfacesWithDifferentConstraints.ts, 54, 1))
 >T : Symbol(T, Decl(twoGenericInterfacesWithDifferentConstraints.ts, 52, 12), Decl(twoGenericInterfacesWithDifferentConstraints.ts, 56, 12))
 

--- a/tests/baselines/reference/twoGenericInterfacesWithDifferentConstraints.types
+++ b/tests/baselines/reference/twoGenericInterfacesWithDifferentConstraints.types
@@ -70,7 +70,7 @@ interface C<T> {
 >x : T
 }
 
-interface C<T extends number> { // error
+interface C<T extends number> { // ok
   y: T;
 >y : T
 }

--- a/tests/cases/compiler/interfaceMergedUnconstrainedNoErrorIrrespectiveOfOrder.ts
+++ b/tests/cases/compiler/interfaceMergedUnconstrainedNoErrorIrrespectiveOfOrder.ts
@@ -1,0 +1,19 @@
+// @filename: working.ts
+// minmal samples from #33395
+export namespace ns {
+    interface Function<T extends (...args: any) => any> {
+        throttle(): Function<T>;
+    }
+    interface Function<T> {
+        unary(): Function<() => ReturnType<T>>;
+    }
+}
+// @filename: regression.ts
+export namespace ns {
+    interface Function<T> {
+        unary(): Function<() => ReturnType<T>>;
+    }
+    interface Function<T extends (...args: any) => any> {
+        throttle(): Function<T>;
+    }
+}

--- a/tests/cases/conformance/interfaces/declarationMerging/twoGenericInterfacesWithDifferentConstraints.ts
+++ b/tests/cases/conformance/interfaces/declarationMerging/twoGenericInterfacesWithDifferentConstraints.ts
@@ -54,7 +54,7 @@ interface C<T> {
   x: T;
 }
 
-interface C<T extends number> { // error
+interface C<T extends number> { // ok
   y: T;
 }
 

